### PR TITLE
Cellery CLI Helm integration.

### PR DESCRIPTION
**Pull request by gnudeep**

### CODE REVIEW

File: components/cli/pkg/commands/setup/create.go [hunk 1]
```diff
@@ -81,20 +81,20 @@ func RunSetupCreateCelleryRuntime(cli cli.Cli, complete bool, isPersistentVolume
...
 		dbUserName = db.DbUserName
 		dbPassword = db.DbPassword
 		if err = cli.Runtime().UpdateNfsServerDetails(nfs.NfsServerIp, nfs.FileShare); err != nil {
 			return fmt.Errorf("failed to update nfs server details")
 		}
 	}
-	if err = cli.Runtime().UpdateMysqlCredentials(dbUserName, dbPassword, dbHostName); err != nil {
-		return fmt.Errorf("error updating mysql credentials: %v", err)
-	}
+	//if err = cli.Runtime().UpdateMysqlCredentials(dbUserName, dbPassword, dbHostName); err != nil {
```
Comment (MadhukaHarith92) [inline components/cli/pkg/commands/setup/create.go]: Shall we just remove these commented lines.

File: components/cli/pkg/runtime/apim.go [hunk 2]
```diff
@@ -19,19 +19,44 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/apim.go]: Formatting issue.

File: components/cli/pkg/runtime/hpa.go [hunk 3]
```diff
@@ -20,6 +20,7 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/hpa.go]: Formattiing issue.

File: components/cli/pkg/runtime/controller.go [hunk 4]
```diff
@@ -19,16 +19,22 @@
...
+	//	if err != nil {
+	//		return err
+	//	}
+	//}
+	runtime.UnmarshalHelmValues("cellery-runtime")
+	runtime.celleryRuntimeVals.Controller.Enabled = true
+	runtime.MarshalHelmValues("cellery-runtime")
+	if err := util.ApplyHelmChartWithCustomValues("cellery-runtime", "cellery-system",
+		"apply", runtime.celleryRuntimeYaml); err != nil {
+		return err
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/controller.go]: It is better if the error could be described (eg : return fmt.Errorf("failed to apply helm charts for controller, %v", err)) and in similar places. Thid eould help to debug. WDYT? But this is not urgent.

File: components/cli/pkg/runtime/idp.go [hunk 5]
```diff
@@ -19,17 +19,30 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/idp.go]: Formatting issue.

File: components/cli/pkg/runtime/istio.go [hunk 6]
```diff
@@ -19,28 +19,35 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/istio.go]: Formatting issue.

File: components/cli/pkg/runtime/knative.go [hunk 7]
```diff
@@ -19,34 +19,42 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/kubernetes"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/knative.go]: Formatting issue.

File: components/cli/pkg/runtime/nginx.go [hunk 8]
```diff
@@ -19,21 +19,36 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/nginx.go]: Formatting issue.

File: components/cli/pkg/runtime/observability.go [hunk 9]
```diff
@@ -19,19 +19,33 @@
 package runtime
 
 import (
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/observability.go]: Formatting issue.

File: components/cli/pkg/runtime/runtime.go [hunk 10]
```diff
@@ -54,33 +56,43 @@ type Runtime interface {
...
 	AddObservability() error
-	AddIdp() error
+	AddIdp(db MysqlDb) error
+	DeleteIdp() error
 	UpdateNodePortIpAddress(nodePortIpAddress string) error
 	CreatePersistentVolume(hasNfs bool) error
 	IsHpaEnabled() (bool, error)
 	WaitFor(checkKnative, hpaEnabled bool) error
 	Validate() error
+	UnmarshalHelmValues(chartName string) error
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/runtime.go]: Shall we rename these 2 functions to something like "SetRuntimeValues", "GetRuntimeValues" to make it more descriptive? WDYT?

File: components/cli/pkg/runtime/runtime.go [hunk 11]
```diff
@@ -338,3 +355,56 @@ func (runtime *CelleryRuntime) Validate() error {
...
+		if err != nil {
+			log.Printf("error: %v", err)
+			return err
+		}
+		runtime.ingressControllerYamls =  string(templateYamls)
+	}
+	return nil
+}
+
+func (runtime *CelleryRuntime) CreateNameSpace(namespace string) error {
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/runtime.go]: There is already an interface function called "CreateNameSpace" in KubeCli interface to create namespace. Shall we use that?

File: components/cli/pkg/util/helm.go [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package util
+
+import (
+	"github.com/cellery-io/sdk/components/cli/pkg/constants"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/util/helm.go]: Formatting issue.

File: components/cli/pkg/util/helm.go [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+	"k8s.io/helm/pkg/renderutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+//Generate k8s artifacts using helm templates.
+func RenderHelmChart(releaseName string, chartNamespace string, chartPath string, chartValues string) string {
```
Comment (MadhukaHarith92) [inline components/cli/pkg/util/helm.go]: Shouldn't this function return an error?

File: components/cli/pkg/util/helm.go [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+	chartTemplate := RenderHelmChart(chartName, namespace, filepath.Join(UserHomeCelleryDir(), constants.HelmCarts, chartName), values)
+	log.Printf(chartTemplate)
+	err := ApplyK8sResource(operation, chartTemplate, namespace)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func CreateNameSpace(namespace string) error {
```
Comment (MadhukaHarith92) [inline components/cli/pkg/util/helm.go]: Ideally, this function should come under kubernetes package and it is already there. Can we reuse it? But this change is not urgent.

**Review by madhukaw — CHANGES_REQUESTED** [review]

File: components/cli/pkg/runtime/hpa.go (hunk 3 above)
Comment (gnudeep) [reply components/cli/pkg/runtime/hpa.go]: Fixed.

File: components/cli/pkg/runtime/idp.go (hunk 5 above)
Comment (gnudeep) [reply components/cli/pkg/runtime/idp.go]: Fixe.

File: components/cli/pkg/runtime/istio.go (hunk 6 above)
Comment (gnudeep) [reply components/cli/pkg/runtime/istio.go]: Fixed.

File: components/cli/pkg/util/helm.go (hunk 12 above)
Comment (gnudeep) [reply components/cli/pkg/util/helm.go]: Fixed.

File: components/cli/pkg/runtime/apim.go (hunk 2 above)
Comment (gnudeep) [reply components/cli/pkg/runtime/apim.go]: Fixe.

File: components/cli/pkg/util/helm.go [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+func ApplyHelmChartWithDefaultValuesCustomCmd(chartName string, namespace string, operation string) error {
+	chartTemplate := RenderHelmChart(chartName, namespace, filepath.Join(UserHomeCelleryDir(), constants.HelmCarts, chartName), "")
+	err := ApplyK8sResource(operation, chartTemplate, namespace)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ApplyHelmChartWithCustomValues(chartName string, namespace string, operation string, values string) error {
```
Comment (MadhukaHarith92) [inline components/cli/pkg/util/helm.go]: IMHO, these functions are related to the runtime and hence should be moved to runtime package. However, this is not an urgent requirement.

**Review by madhukaw — APPROVED** [review]

File: components/cli/pkg/runtime/runtime.go (hunk 10 above)
Comment (gnudeep) [reply components/cli/pkg/runtime/runtime.go]: +1 for the change.

File: components/cli/pkg/runtime/hpa.go [hunk 16]
```diff
@@ -23,6 +23,8 @@ import (
 	"fmt"
 	"strings"
 
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/hpa.go]: Code formatting issue.

**Review by madhukaw — CHANGES_REQUESTED** [review]

File: components/cli/pkg/runtime/nginx.go [hunk 17]
```diff
@@ -21,19 +21,35 @@ package runtime
 import (
 	"encoding/json"
 	"fmt"
+	"log"
 	"path/filepath"
 
+	"cellery.io/cellery/components/cli/pkg/util"
```
Comment (MadhukaHarith92) [inline components/cli/pkg/runtime/nginx.go]: Code formatting issue.

**Pull request closed without merging**